k8s/src/main/scala/io/buoyant/k8s/istio/IdentifierPreconditions.scala

```diff
@@ -23,32 +33,51 @@ trait IdentifierPreconditions {
     }
   }
 
-  def matchesAllConditions(getHeader: (String) => Option[String], matchCondition: MatchCondition): Boolean = {
+  def matchesAllConditions(req: IstioRequestMeta, matchCondition: MatchCondition): Boolean = {
     val matchesHeaders = matchCondition.`httpHeaders`.forall {
+      case ("uri", m) => headerMatches(req.uri, m)
+      case ("scheme", m) => headerMatches(req.scheme, m)
+      case ("method", m) => headerMatches(req.method, m)
+      case ("authority", m) => headerMatches(req.authority, m)
       case (headerName, stringMatch) =>
-        // return false if the headerName does not appear in the request's headerMap
-        getHeader(headerName) match {
-          case Some(headerValue) => headerMatches(headerValue, stringMatch)
+        req.getHeader(headerName) match {
+          case Some(a) => headerMatches(a, stringMatch)
           case None => false
         }
     }
     //TODO: add other match conditions
     matchesHeaders
   }
 
-  def filterRules(rules: Map[String, RouteRule], dest: String, getHeader: (String) => Option[String]): Seq[(String, RouteRule)] = rules.filter {
+  def httpRewrite(rule: RouteRule, uri: String, authority: Option[String]): (String, Option[String]) = {
+    rule.`rewrite` match {
+      case Some(HTTPRewrite(url, updatedAuth)) =>
+        val updatedUri = url.map { replacement =>
+          rule.`match`.flatMap(_.`httpHeaders`.get("uri").flatMap(_.`matchType`)) match {
+            case Some(OneofMatchType.Prefix(pfx)) =>
+              uri.replace(pfx, replacement)
+            case _ =>
+              uri.split("/").head + replacement
+          }
+        }
+        (updatedUri.getOrElse(uri), updatedAuth.orElse(authority))
+      case _ => (uri, authority)
+    }
+  }
+
+  def filterRules(rules: Map[String, RouteRule], dest: String, req: IstioRequestMeta): Seq[(String, RouteRule)] = rules.filter {
     case (_, r) if r.`destination`.contains(dest) =>
       // return true if no match conditions were defined on the route-rule
-      r.`match`.forall(matchesAllConditions(getHeader, _))
+      r.`match`.forall(matchesAllConditions(req, _))
     case _ => false
   }.toSeq
 
-  def maxPrecedenceRuleName(rules: Seq[(String, RouteRule)]): Option[String] = {
+  def maxPrecedenceRuleName(rules: Seq[(String, RouteRule)]): Option[(String, RouteRule)] = {
```

Comment:
We should probably change the name of this method now that it's no longer just returning the name.